

```bash
#TODO: assembly checkpoints
#TODO: checkm checkpoint after dereplication (it takes ~7 minutes), if checkm is run in bg we could talk about drep tables in the mean time
#genomad checkpoint

```

# Pacbio HiFi

If this is not the case yet remember to activate the correct conda env:

    conda activate LongReads

## Dataset
We are going to use a PacBio HiFi Zymo mock community, available here:

    ~/data/public/teachdata/ebame-2022/metagenomics/HIFI_datasets/samples/Zymo_sample1e5.fastq.gz

As previously we can use the command seqkit stats to assess this sample.
<details><summary>Solution</summary>
<p>

```bash
seqkit stats ~/data/public/teachdata/ebame-2022/metagenomics/HIFI_datasets/samples/Zymo_sample1e5.fastq.gz
```

</p>
</details>

 ## Assembly


Mostly 3 software have been developed for assembling HiFi metagenomic datasets.

- [metaFlye](https://www.nature.com/articles/s41592-020-00971-x) 

- [hifiasm-meta](https://www.nature.com/articles/s41592-022-01478-3)

- [metaMDBG](https://www.biorxiv.org/content/10.1101/2023.07.07.548136v1)

In this tutorial, we are going to run hifiasm-meta and metaMDBG separatly, then use a method to reconciliate their results.

First be sure to work in this directory:

```bash
mkdir -p ~/data/mydatalocal/HiFi
cd ~/data/mydatalocal/HiFi
```

Let's create directories for the assemblies:

```bash
mkdir -p ~/data/mydatalocal/HiFi/assemblies/
mkdir -p ~/data/mydatalocal/HiFi/assemblies/hifiasm_meta/
mkdir -p ~/data/mydatalocal/HiFi/assemblies/metaMDBG/
```

As usual, try to craft your own command line to run the assemblers. 
Use the following commands to see usage information:
```bash
hifiasm_meta -h
metaMDBG asm -h
```

<details><summary>Solution</summary>
<p>

```bash
cd ~/data/mydatalocal/HiFi
hifiasm_meta -o ~/data/mydatalocal/HiFi/assemblies/hifiasm_meta/asm ~/data/public/teachdata/ebame-2022/metagenomics/HIFI_datasets/samples/Zymo_sample1e5.fastq.gz -t 4
metaMDBG asm ~/data/mydatalocal/HiFi/assemblies/metaMDBG/ ~/data/public/teachdata/ebame-2022/metagenomics/HIFI_datasets/samples/Zymo_sample1e5.fastq.gz -t 4
```
</p>
</details>

This should take about 4858.465 seconds

So instead lets comment on the pre-run version.
```bash
cd ~/data/mydatalocal/HiFi
ln -s ~/data/public/teachdata/ebame-2022/metagenomics/HIFI_datasets/HumanReal_asm prerun_asm
```

Here assembly is also available in the form of a .gfa. Hifiasm uses the GFA1  [format](http://gfa-spec.github.io/GFA-spec/GFA1.html). but also adds on entry relative to reads (line starting by A). 

There is a quite a diversity of output file, what are they? What is a unitig a contig? Let's check the [documentation](https://hifiasm.readthedocs.io/en/latest/interpreting-output.html)

Let's use Bandage to assess this assembly.
```bash
cd ~/data/mydatalocal/HiFi/prerun_asm
Bandage load asm.p_ctg.gfa
```

<details><summary> If you have issues with Bandage </summary>
<p>

#### Fix1
Did you use -X or -Y when connecting to the VM? If not, please disconect and retype ssh with that flag:

    ssh -X ubuntu@xxx.xxx.xxx.xxx

#### Fix2
If you have Bandage on your laptop, use the scp command to download the gfa file on your laptop:

    scp ubuntu@xxx.xxx.xxx.xxx:~/data/mydatalocal/HiFi/prerun_asm/asm.p_ctg.gfa	.

This will copy the file to the directory you executed that command from. Also to be clear this command should not be run on the vm. This is a command for your laptop to request that file from the distant server. So it should be run on a terminal before you connect to the vm.

#### Fix3
Try and follow explanation on how to forward display from this google doc:
https://docs.google.com/document/d/1VPnL-5mXXQimkXQNiQagPhgzRn8j1JBHCLV42r8-Wqc/edit#

</p>
</details>


-->  look at assembly statistics

-->  look at actual size of smaller size contigs

--> What are the circular components?


The Zymo is a bit more exciting than the HumanReal in terms of circular components, however some of those long contigs even if not circular are could already satisfy medium or high MAGs criterion for quality.

 ## MAGs?
Let's focus here only on circular contigs. 

#### Extract circular contigs

First, let's transform hifiasm-meta outputs in a conventional fasta format:
```bash
awk '/^S/{print ">"$2;print $3}' ~/data/mydatalocal/HiFi/assemblies/hifiasm_meta/asm.p_ctg.gfa > ~/data/mydatalocal/HiFi/assemblies/hifiasm_meta/asm.p_ctg.fasta
```

We can know if a contig is circular by looking at contig headers in the fasta files (the line starting with a ">").
If a header ends with a "c", it means that the contig is circular, otherwise it is linear. You can check this info with the following command:

```bash
cat ~/data/mydatalocal/HiFi/assemblies/hifiasm_meta/asm.p_ctg.fasta | grep ">"
```

Let's create folders for the circular contigs:
```bash
mkdir -p ~/data/mydatalocal/HiFi/circularContigs/
mkdir -p ~/data/mydatalocal/HiFi/circularContigs/hifiasm_meta/
mkdir -p ~/data/mydatalocal/HiFi/circularContigs/metaMDBG/
```

Now, try to run the following homemade script to extract the circular contigs:

    ~/repos/Ebame/scripts/extractCircularContigs.py 

<details><summary>It should not be too hard if you use the -h</summary>
<p>

```bash
~/repos/Ebame/scripts/extractCircularContigs.py ~/data/mydatalocal/HiFi/assemblies/hifiasm_meta/asm.p_ctg.fasta ~/data/mydatalocal/HiFi/circularContigs/hifiasm-meta/
~/repos/Ebame/scripts/extractCircularContigs.py ~/data/mydatalocal/HiFi/assemblies/metaMDBG/contigs.fasta.gz ~/data/mydatalocal/HiFi/circularContigs/metaMDBG/
```

</p>
</details>

#### Assembly reconciliation

We are now going to merge the results of metaMDBG and hifiasm-meta. The idea is to compute the similarity (ANI) between the circular contigs, and to choose only one represententive if two contigs are duplicated (ANI > 0.95 by default).

For this task, we are going to use the software dRep. dRep has a lot of options, let's craft the command together, first display de-replicate options:

```bash
dRep dereplicate -h
```

<details><summary>dRep takes a list of genomes as input (option -g), let's add all circular contig paths in a single file:</summary>
<p>

```bash
#Collect circular contig paths
ls ~/data/mydatalocal/HiFi/circularContigs/hifiasm-meta/*.fa > ~/data/mydatalocal/HiFi/circularContigs/allCircularContigs.txt
ls ~/data/mydatalocal/HiFi/circularContigs/metaMDBG/c*.fa >> ~/data/mydatalocal/HiFi/circularContigs/allCircularContigs.txt

#Check input file
cat ~/data/mydatalocal/HiFi/circularContigs/allCircularContigs.txt
```

</p>
</details>

<details><summary>We are going to use dRep in a quick way, first disable quality check with option --ignoreGenomeQuality (we'll do it after dereplication ), and select the "fastANI" method for comparing the circular contigs:</summary>
<p>

```bash
dRep dereplicate ~/data/mydatalocal/HiFi/circularContigs/drep/ -p 4 -g ~/data/mydatalocal/HiFi/circularContigs/allCircularContigs.txt --S_algorithm fastANI --ignoreGenomeQuality
```

</p>
</details>

The list of dereplicated circular contigs is located here:
```bash
ls -lh ~/data/mydatalocal/HiFi/circularContigs/drep/dereplicated_genomes/
```

dRep provides a lot of useful information, for instance, we can look at the similarity between the pair of circular contigs:
```bash
column -s, -t < ~/data/mydatalocal/HiFi/circularContigs/drep/data_tables/Ndb.csv
```

Are there any circular contigs which are only found by one assembler?
```bash
column -s, -t < ~/data/mydatalocal/HiFi/circularContigs/drep/data_tables/
```

#### Check circular contigs
Clearly some of these are not genomes, let's run checkm on the dereplicated contigs:

```bash
checkm lineage_wf ~/data/mydatalocal/HiFi/circularContigs/drep/dereplicated_genomes/ ~/data/mydatalocal/HiFi/circularContigs/drep/dereplicated_genomes/checkm/ -r -x .fa -t 4 --pplacer_threads 4 --tab_table
```

Check columns corresponding to completeness and contamination:

```bash
awk -F"\t" '{ print $1, "\t", $12, "\t", $13 }' ~/data/mydatalocal/HiFi/circularContigs/drep/dereplicated_genomes/checkm/results.tsv
```

## Plasmids and virus?

Some of these smaller contigs are likely to be plasmids or virus. Let's use [genomad](https://github.com/apcamargo/genomad), a machine learning approach to verify this.

As usual, let's check the software usage information:
```bash
genomad -h
genomad end-to-end -h
```

<details><summary>Genomad takes as input a single fasta file. It will then process the contigs one by one and determine how likely they are to be plasmids or virus. Let's concatenate the small dereplicated circular contigs in a single fasta file. </summary>
<p>

```bash
#Concatenate all circular contigs in a single fasta file
cat ~/data/mydatalocal/HiFi/circularContigs/drep/dereplicated_genomes/*.fa > ~/data/mydatalocal/HiFi/circularContigs/allCircularContigs.fasta

#Concatenante only small circular contigs
find ~/data/mydatalocal/HiFi/circularContigs/drep/dereplicated_genomes/*.fa -size -500k | xargs cat > ~/data/mydatalocal/HiFi/circularContigs/allSmallCircularContigs.fasta
```

</p>
</details> 

The genomad database is located here:

    ~/data/public/teachdata/ebame-2023/virome/db/genomad_db
    
<details><summary>Let's run genomad (you can use option --sensitivity 1.0 to speed-up prediction, use only for this tutorial):</summary>
<p>

```bash
genomad end-to-end ~/data/mydatalocal/HiFi/circularContigs/allSmallCircularContigs.fasta ~/data/mydatalocal/HiFi/circularContigs/genomad/ ~/data/public/teachdata/ebame-2023/virome/db/genomad_db --threads 4 --sensitivity 1.0
```

</p>
</details> 

</p>
</details> 

<details><summary>Read genomad logs and try to print plasmids and virus summaries:</summary>
<p>

```bash
cat ~/data/mydatalocal/HiFi/circularContigs/genomad_s1/allSmallCircularContigs_summary/allSmallCircularContigs_plasmid_summary.tsv
cat ~/data/mydatalocal/HiFi/circularContigs/genomad_s1/allSmallCircularContigs_summary/allSmallCircularContigs_virus_summary.tsv
```

</p>
</details> 

## Other contigs

#### But wait what if I want to look at one of the non circular contigs?
Retry to use checkm on a contigs you chose and saved with Bandage.

--> find a long contigs from Bandage. Copy the name
--> on the server, use the following command line replacing \<NODE\>:

```bash
cd ~/data/mydatalocal/HiFi
mkdir linear_contigs
Bandage reduce  prerun_asm/asm.p_ctg.gfa linear_contigs/<Node>.gfa  --scope aroundnodes --nodes <NODE> --distance 0
```
--> use a bash online to extract, name and sequence from that gfa graph:
```bash
cd ~/data/mydatalocal/HiFi/linear_contigs
awk '/^S/{print ">"$2"\n"$3}' <Node>.gfa > <Node>.fa

```
--> use checkm 